next: Parse bracket loops at the statement level

dlongnecke-cray opened the pull request:
next: Parse bracket loops at the statement level (#17822)

This is for `compiler/next`.

A "bracket loop" is a loop construct that looks like:

```chapel
[i in 0..15] writeln(i);
```

A bracket loop is different from a forall loop in that it will try to
use a serial iterator if a parallel iterator is not available.

Add the header `BracketLoop`, and parse bracket loops that appear at
a statement level.

Rename `usesDo` to `usesImplicitBlock`. Remove `isBodyBlock` from
the `DoWhile` and `BracketLoop` nodes. Rewire the parser accordingly.

Clean up parser actions for loops by adding helpers to `ParserContext`.

Reviewed by @lydia-duncan and @mppf. Thanks!

TESTING:

- [x] All `compiler/next` tests pass

Signed-off-by: David Longnecker <dlongnecke-cray@users.noreply.github.com>

dlongnecke-cray commented:
@mmpf Could you do a quick pass over this? I am concerned about `isBodyBlock()` and wonder if it is somewhat redundant with `usesDo`. I am wondering if we can perhaps merge the two together. E.g. for other loops where `usesDo` can be true, we could instead replace that with `isBodyBlock` such that:

```cpp
if (isBodyBlock() && onlyStatementIsBlock) // Then the construct must have used do...
```

Or maybe it's nothing to be worried about? I am not sure how many constructs use the `stmt` production instead of `do_stmt`.

dlongnecke-cray commented:
Another thing I am worried about is whether we should have some "build" functions in `ParserContextImpl` in order to clean up all the redundant parser actions. I am not sure how much we care about reducing duplication in the parser, though.

dlongnecke-cray commented:
@vasslitvinov Would you be the primary reviewer for this?

mppf commented:
> Could you do a quick pass over this @mppf? I am concerned about isBodyBlock() and wonder if it is somewhat redundant with usesDo. I am wondering if we can perhaps merge the two together. E.g. for other loops where usesDo can be true, we could instead replace that with isBodyBlock such that:

I think that `usesDo` is just fine, but you were arguing it's not a great name for it in the context of do/while loops. So, I think the concept of `usesDo` is fine but the remedy is to just rename the function. We could rename the function to `usesImplicitBlock` for example. I don't think we need to be *too* fussy about this name because it will be very rarely used. Anyway, I think that whatever it is called, it should be in `Loop`.

> Another thing I am worried about is whether we should have some "build" functions in ParserContextImpl in order to clean up all the redundant parser actions. I am not sure how much we care about reducing duplication in the parser, though.

Yes, we have already added a few of these (e.g. `buildBinOp`, `buildFunctionDecl`). I think it is worth doing when we can reduce redundant code.

dlongnecke-cray commented:
Ok, I've gone ahead and renamed `usesDo` to `usesImplicitBlock`, and removed `isBodyBlock` for `DoWhile` and `BracketLoop`. I've also added helpers to `ParserContext` to clean up parser actions.

lydia-duncan commented on compiler/next/include/chpl/uast/BracketLoop.h:
```diff
@@ -0,0 +1,92 @@
+         int loopBodyChildNum,
+         int numLoopBodyStmts,
+         bool usesImplicitBlock,
+         bool isExpressionLevel)
```
The indentation on these arguments is a little wonky.  Not a big deal, just pointing it out

dlongnecke-cray replied:
Fixed, thanks!

lydia-duncan reviewed (approved):
Just a minor nit, looks good!
